File: core/yohu-domain/src/log_filter.rs
//! 日志过滤匹配（ADR-v6-006）：wire [`LogFilter`] 的领域语义。
//!
//! protocol 只持有结构；导出与 UI `matchesWireFilter` 共用本层 + testdata/log_filter.json。回补读环不过滤。
//! Tag：`tag_contains` 按逗号 / 分号 / `|` 拆多针，任一 OrdinalIgnoreCase **精确**命中；空或仅分隔符 = 不限。

use yohu_protocol::{LogFilter, LogLine, LogScope};

/// 级别序：V < D < I < W < E < F；未知为 0（低于任意已知级别）。
pub fn level_rank(level: char) -> u8 {
    match level {
        'V' | 'v' => 1,
        'D' | 'd' => 2,
        'I' | 'i' => 3,
        'W' | 'w' => 4,
        'E' | 'e' => 5,
        'F' | 'f' => 6,
        _ => 0,
    }
}

/// ASCII 忽略大小写的子串匹配（ADR-v6-006：关键字 = OrdinalIgnoreCase 包含）。
/// 用字节级 `eq_ignore_ascii_case`，避免 Unicode `to_lowercase()` 对非 ASCII 的语义偏差。
fn contains_ascii_ignore_case(haystack: &str, needle: &str) -> bool {
    let h = haystack.as_bytes();
    let n = needle.as_bytes();
    if n.is_empty() {
        return true;
    }
    if n.len() > h.len() {
        return false;
    }
    h.windows(n.len())
        .any(|w| w.iter().zip(n).all(|(a, b)| a.eq_ignore_ascii_case(b)))
}

fn level_in_set(levels: &[char], line: char) -> bool {
    if levels.is_empty() {
        return true;
    }
    let needle = line.to_ascii_uppercase();
    levels.iter().any(|item| item.to_ascii_uppercase() == needle)
}

fn is_tag_needle_sep(c: char) -> bool {
    matches!(c, ',' | '，' | '、' | ';' | '；' | '|')
}

/// 拆 `tag_contains`；空段丢掉。空结果 = 不限 Tag。空白留在针内。
fn tag_needles(spec: &str) -> impl Iterator<Item = &str> {
    spec.split(is_tag_needle_sep)
        .map(str::trim)
        .filter(|part| !part.is_empty())
}

fn equals_ascii_ignore_case(a: &str, b: &str) -> bool {
    a.len() == b.len()
        && a.as_bytes()
            .iter()
            .zip(b.as_bytes())
            .all(|(x, y)| x.eq_ignore_ascii_case(y))
}

fn tag_allowed(line_tag: &str, spec: &str) -> bool {
    let mut needles = tag_needles(spec).peekable();
    if needles.peek().is_none() {
        return true;
    }
    needles.any(|needle| equals_ascii_ignore_case(line_tag, needle))
}

/// 单行匹配。`Package { pids: [] }` 不命中任何行。
/// 级别：`levels` 空则不限；非空则精确属于该集合（不是最低含以上）。
/// Tag：多针 OR；关键字仍是整段包含。
pub fn log_filter_matches(filter: &LogFilter, line: &LogLine) -> bool {
    if !level_in_set(&filter.levels, line.level) {
        return false;
    }
    if let Some(tag) = &filter.tag_contains {
        if !tag_allowed(&line.tag, tag) {
            return false;
        }
    }
    if let Some(msg) = &filter.message_contains {
        if !contains_ascii_ignore_case(&line.msg, msg) {
            return false;
        }
    }
    match &filter.scope {
        LogScope::All => true,
        LogScope::Pid { pid } => line.pid == *pid,
        LogScope::Package { pids } => pids.contains(&line.pid),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use yohu_protocol::LogLine;

    #[test]
    fn level_rank_order() {
        assert!(level_rank('V') < level_rank('D'));
        assert!(level_rank('D') < level_rank('I'));
        assert!(level_rank('I') < level_rank('W'));
        assert!(level_rank('W') < level_rank('E'));
        assert!(level_rank('E') < level_rank('F'));
        assert_eq!(level_rank('?'), 0);
    }

    #[test]
    fn matches_shared_fixture() {
        #[derive(serde::Deserialize)]
        struct Case {
            line: LogLine,
            filter: LogFilter,
            expect: bool,
        }
        let cases: Vec<Case> =
            serde_json::from_str(include_str!("../testdata/log_filter.json")).expect("fixture");
        for (i, case) in cases.iter().enumerate() {
            assert_eq!(
                log_filter_matches(&case.filter, &case.line),
                case.expect,
                "case {i}"
            );
        }
    }

    #[test]
    fn level_rank_shared_fixture() {
        #[derive(serde::Deserialize)]
        struct Case {
            level: char,
            rank: u8,
        }
        let cases: Vec<Case> =
            serde_json::from_str(include_str!("../testdata/level_rank.json")).expect("fixture");
        for (i, case) in cases.iter().enumerate() {
            assert_eq!(level_rank(case.level), case.rank, "case {i}");
        }
    }

    #[test]
    fn known_levels_shared_fixture() {
        let letters: Vec<char> =
            serde_json::from_str(include_str!("../testdata/log_levels.json")).expect("fixture");
        assert_eq!(letters, vec!['V', 'D', 'I', 'W', 'E', 'F']);
        for (i, letter) in letters.iter().enumerate() {
            assert_eq!(level_rank(*letter), (i + 1) as u8, "letter {letter}");
        }
    }
}

File: ui/packages/api/src/log-filter.ts
/**
 * 日志过滤匹配。与 yohu-domain::log_filter 同一套 testdata/log_filter.json。
 * 级别字母单源 testdata/log_levels.json ↔ LEVELS。
 */

import type { LogFilter, LogLine, LogScope } from "./types";

export const LEVELS = ["V", "D", "I", "W", "E", "F"] as const;
export type LevelLetter = (typeof LEVELS)[number];

export function parseLevelLetter(level: string): LevelLetter | null {
  const upper = level.toUpperCase();
  return (LEVELS as readonly string[]).includes(upper) && upper.length === 1 ? (upper as LevelLetter) : null;
}

export function isLogLevelLetter(token: string): boolean {
  return parseLevelLetter(token) !== null;
}

/** 只保留 LEVELS，去重并按 V→F。空 = 不限级别。 */
export function normalizeLevels(input: readonly string[]): LevelLetter[] {
  const seen = new Set<LevelLetter>();
  for (const raw of input) {
    const letter = parseLevelLetter(raw);
    if (letter) seen.add(letter);
  }
  return LEVELS.filter((letter) => seen.has(letter));
}

export function parseTagNeedles(raw: string): string[] {
  if (!raw) return [];
  const needles: string[] = [];
  for (const part of raw.split(/[,，、;；|]/)) {
    const needle = part.trim();
    if (needle) needles.push(needle);
  }
  return needles;
}

export function tagFilterActive(raw: string): boolean {
  return parseTagNeedles(raw).length > 0;
}

/** ASCII 忽略大小写精确等价（已提交 Tag 针，不是子串；避免 libc 命中 libcomposer_ext）。 */
export function equalsAsciiIgnoreCase(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    const x = a.charCodeAt(i);
    const y = b.charCodeAt(i);
    if (x === y) continue;
    const xl = x >= 65 && x <= 90 ? x + 32 : x;
    const yl = y >= 65 && y <= 90 ? y + 32 : y;
    if (xl !== yl) return false;
  }
  return true;
}

/** ASCII 忽略大小写前缀匹配（输入中草稿 Tag 针，如 GGEC- 命中 GGEC-DeviceEventRouter）。 */
export function startsWithAsciiIgnoreCase(haystack: string, prefix: string): boolean {
  if (prefix.length > haystack.length) return false;
  for (let i = 0; i < prefix.length; i++) {
    const x = haystack.charCodeAt(i);
    const y = prefix.charCodeAt(i);
    if (x === y) continue;
    const xl = x >= 65 && x <= 90 ? x + 32 : x;
    const yl = y >= 65 && y <= 90 ? y + 32 : y;
    if (xl !== yl) return false;
  }
  return true;
}

/** ASCII 仅 A–Z 折叠后的子串起点；空针同 `String.indexOf`。 */
export function indexOfAsciiIgnoreCase(haystack: string, needle: string, from = 0): number {
  if (needle.length === 0) {
    if (from < 0) return 0;
    return from <= haystack.length ? from : haystack.length;
  }
  const start = from < 0 ? 0 : from;
  const n = needle.length;
  const last = haystack.length - n;
  outer: for (let i = start; i <= last; i++) {
    for (let j = 0; j < n; j++) {
      const a = haystack.charCodeAt(i + j);
      const b = needle.charCodeAt(j);
      if (a === b) continue;
      const al = a >= 65 && a <= 90 ? a + 32 : a;
      const bl = b >= 65 && b <= 90 ? b + 32 : b;
      if (al !== bl) continue outer;
    }
    return i;
  }
  return -1;
}

const TAG_TRAILING_SEP = /[,，、;；|]\s*$/;

/** 拆解为已提交针（精确匹配）与当前草稿针（前缀匹配；尾部有分隔符则无草稿）。 */
export function parseTagFilterNeedles(spec: string): { committed: string[]; draft: string } {
  const needles = parseTagNeedles(spec);
  if (needles.length === 0) return { committed: [], draft: "" };
  if (TAG_TRAILING_SEP.test(spec)) return { committed: needles, draft: "" };
  return { committed: needles.slice(0, -1), draft: needles[needles.length - 1]! };
}

/** 空针不限；已提交针精确命中，正在输入的草稿针前缀命中（打字实时匹配且避免 libc 误伤）。 */
export function tagAllowed(lineTag: string, spec: string): boolean {
  const { committed, draft } = parseTagFilterNeedles(spec);
  if (committed.length === 0 && !draft) return true;
  for (const needle of committed) {
    if (equalsAsciiIgnoreCase(lineTag, needle)) return true;
  }
  if (draft && startsWithAsciiIgnoreCase(lineTag, draft)) {
    return true;
  }
  return false;
}

export function containsAsciiIgnoreCase(haystack: string, needle: string): boolean {
  return indexOfAsciiIgnoreCase(haystack, needle) >= 0;
}

export function levelAllowed(lineLevel: string, selected: readonly string[]): boolean {
  if (selected.length === 0) return true;
  const letter = parseLevelLetter(lineLevel);
  if (!letter) return false;
  for (const item of selected) {
    if (parseLevelLetter(item) === letter) return true;
  }
  return false;
}
function scopeMatches(line: LogLine, scope: LogScope): boolean {
  switch (scope.kind) {
    case "all":
      return true;
    case "pid":
      return line.pid === scope.pid;
    case "package":
      return scope.pids.includes(line.pid);
  }
}

/** 与 yohu-domain::log_filter_matches 同一语义。 */
export function logFilterMatches(filter: LogFilter, line: LogLine): boolean {
  if (!levelAllowed(line.level, filter.levels ?? [])) return false;
  if (filter.tag_contains !== undefined && !tagAllowed(line.tag, filter.tag_contains)) return false;
  if (filter.message_contains && !containsAsciiIgnoreCase(line.msg, filter.message_contains)) {
    return false;
  }
  return scopeMatches(line, filter.scope);
}

export function matchesWireFilter(line: LogLine, filter: LogFilter): boolean {
  return logFilterMatches(filter, line);
}
